src/utils/instrumentation.ts

```diff
@@ -0,0 +1,201 @@
+/*!
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import arrify = require('arrify');
+import path = require('path');
+import {google} from '../../protos/protos';
+import {Entry} from '../entry';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+declare const global: {[index: string]: any};
+
+// The global variable keeping track if instrumentation record was already written or not.
+// The instrumentation record should be generated only once per process and contain logging
+// libraries related info.
+global.instrumentationAdded = false;
+
+// The variable to hold cached library version
+let libraryVersion: string;
+
+// Max length for instrumentation library name and version values
+const maxDiagnosticValueLen = 14;
+
+export const DIAGNOSTIC_INFO_KEY = 'logging.googleapis.com/diagnostic';
+export const INSTRUMENTATION_SOURCE_KEY = 'instrumentation_source';
+export const NODEJS_LIBRARY_NAME_PREFIX = 'nodejs';
+export type InstrumentationInfo = {name: string; version: string};
+
+/**
+ * This method returns the status if instrumentation info was already added or not.
+ * @returns true if the log record with instrumentation info was already added, false otherwise.
+ */
+export function getInstrumentationInfoStatus() {
+  return global.instrumentationAdded;
+}
+
+/**
+ * This method helps to populate entries with instrumentation data
+ * @param entry {Entry} The entry or array of entries to be populated with instrumentation info
+ * @returns {Entry} Array of entries which contains an entry with current library instrumentation info
+ */
+export function populatedInstrumentationInfo(entry: Entry | Entry[]): Entry[] {
+  // Update the flag indicating that instrumentation entry was already added once,
+  // so any subsequent calls to this method will not add a separate instrumentation log entry
+  let isWritten = global.instrumentationAdded;
+  global.instrumentationAdded = true;
```

Comment:
nit: if the ingestion fails does the flag remains set?

Reply:
Yes, it remains - we rely on retries to deliver an entry. Will look to see if I can track the request with callback, however it might be dangerous - what if error still returned for request containing oversized entries despite a fact that some entries were actually committed?

Reply:
I checked and indeed, the error for entire request will be returned in case when some of entries were dropped... We probable should address this in separate PR - it requires deeper understanding and for now I believe retries should do the job.